clippy_lints/src/print.rs

```diff
@@ -135,6 +130,61 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
     }
 }
 
+// Check for print!("... \n", ...).
+fn check_print<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, span: Span, args: &HirVec<Expr>) {
+    if_let_chain!{[
+        // ensure we're calling Arguments::new_v1
+        args.len() == 1,
+        let ExprCall(ref args_fun, ref args_args) = args[0].node,
+        let ExprPath(ref qpath) = args_fun.node,
+        let Some(const_def_id) = opt_def_id(resolve_node(cx, qpath, args_fun.hir_id)),
+        match_def_path(cx.tcx, const_def_id, &paths::FMT_ARGUMENTS_NEWV1),
+        args_args.len() == 2,
+        let ExprAddrOf(_, ref match_expr) = args_args[1].node,
+        let ExprMatch(ref args, _, _) = match_expr.node,
+        let ExprTup(ref args) = args.node,
+
+        // collect the format string parts and check the last one
+        let Some((fmtstr, fmtlen)) = get_argument_fmtstr_parts(&args_args[0]),
+        let Some('\n') = fmtstr.chars().last(),
+
+        // "foo{}bar" is made into two strings + one argument,
+        // if the format string starts with `{}` (eg. "{}foo"),
+        // the string array is prepended an empty string "".
+        // We only want to check the last string after any `{}`:
+        args.len() < fmtlen,
+    ], {
+        span_lint(cx, PRINT_WITH_NEWLINE, span,
+                  "using `print!()` with a format string that ends in a \
+                   newline, consider using `println!()` instead");
+    }}
+}
+
+/// Check for println!("")
+fn check_println<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, span: Span, args: &HirVec<Expr>) {
+    if_let_chain!{[
+        // ensure we're calling Arguments::new_v1
+        args.len() == 1,
+        let ExprCall(ref args_fun, ref args_args) = args[0].node,
+        let ExprPath(ref qpath) = args_fun.node,
+        let Some(const_def_id) = opt_def_id(resolve_node(cx, qpath, args_fun.hir_id)),
+        match_def_path(cx.tcx, const_def_id, &paths::FMT_ARGUMENTS_NEWV1),
+        args_args.len() == 2,
+
+        // check that the string is empty
+        let Some((fmtstr, 1)) = get_argument_fmtstr_parts(&args_args[0]),
+        fmtstr.deref() == "\n",
+
+        // HACK: check size of the macro invocation:
+        // println!();  // 11 chars: don't lint
+        // println!("") // 12 chars: lint
+        (span.hi() - span.lo()).0 > 11,
```

Comment:
hahahaha! That is awesome, but I think it would be better if you checked whether the code snippet matches `println!("")`